DOCKER-793: First pass for NFS shared volumes

misterdjules opened the pull request:
This is a first pass that allows joyent/sdc-docker to provide [experimental support for NFS shared  modules](https://github.com/joyent/sdc-docker/pull/78). My plan is to add and update tests when the volumes API stabilizes a bit.

trentm commented on lib/volapi.js:
```diff
@@ -0,0 +1,150 @@
+};
+
+VOLAPI.prototype.createVolume =
+    function createVolume(params, options, callback) {
```
Hrm. I'd like to propose that some of us talk (tomorrow?) about picking a preference for function signature for clients, at least for new ones. My preference has moved to wanting `function theName(opts, cb)` for all methods. Here I mean that there are two arguments: the first is options, the second is the callback. I notice that we aren't that consistent (surprise surprise!).

(A separate less important discussion is the what to call those vars: `opts` vs `options`, `callback` vs `cb`.)

trentm replied:
Also discussable is whether all client methods should be strict in validating input options. I'm leaning towards *yes*.

misterdjules replied:
> Also discussable is whether all client methods should be strict in validating input options. I'm leaning towards yes.

Do you mean strictly validating the properties of the `options` object passed to `createVolume` and all other exposed methods?

trentm replied:
Yes, that's what I meant. E.g. https://github.com/joyent/sdcadm/blob/f7e0a0eeedb8666032a3151caa9a2a31736ae3fe/lib/common.js#L40-L75

trentm commented on lib/volapi.js:
```diff
@@ -0,0 +1,150 @@
+VOLAPI.prototype.createVolume =
+    function createVolume(params, options, callback) {
+        if (typeof (options) === 'function') {
+            callback = options;
```
I'm not a fan of this double-indentation style just because the function name and/or args was long. That width for just the start of the method should impact indentation for all the lines of the function,IMHO.

misterdjules replied:
Agreed, will fix.

trentm commented on lib/volapi.js:
```diff
@@ -0,0 +1,150 @@
+var util = require('util');
```
license and copyright block missing?

Please have two blocks of requires: external packages and then local libs.

misterdjules replied:
Sounds good, will fix.

trentm commented on lib/volapi.js:
```diff
@@ -0,0 +1,150 @@
+function VOLAPI(options) {
+    RestifyClient.call(this, options);
+    this.url = options.url;
+}
```
I'd love to move to requiring user-agent in our clients. That could be a separate PR, tho.

trentm replied:
Might consider requiring specifying an API version -- even if volapi doesn't do API versioning yet. Require users to specify "acceptVersion: ~1" or explicitly "acceptVersion: *" and it makes moving to api versioning in volapi much easier later.

misterdjules replied:
Sounds good, will fix.

trentm commented on package.json:
```diff
@@ -1,7 +1,7 @@
-  "version": "9.2.0",
+  "version": "9.3.0",
   "homepage": "http://www.joyent.com",
   "repository": {
```
Should update assert-plus to `1.*` latest, it has fixes for a number of methods, esp. the "optional*" ones.

misterdjules replied:
Good point, will do.

trentm commented on lib/volapi.js:
```diff
@@ -0,0 +1,150 @@
+        }
+
+        if (params.context) {
+            reqOpts.headers['x-context'] = JSON.stringify(params.context);
```
Do we know that this is needed and gets carried through? Assuming this is for either MachineAudit in cloudapi and/or for RBAC?

misterdjules replied:
I assumed this was used by RBAC, and saw it was setup in the same way in the APIs for other services, but I'll double check.

trentm commented on test/volapi.test.js:
```diff
@@ -0,0 +1,62 @@
+ * Copyright (c) 2016, Joyent, Inc.
+ */
+
+var VOLAPI = require('../lib/index').VOLAPI;
```
Local lib imports in a separate block below the external package imports.

misterdjules replied:
Will fix.

trentm commented on test/volapi.test.js:
```diff
@@ -0,0 +1,62 @@
+
+exports.setUp = function (callback) {
+    var logger = new bunyan.createLogger({
+            name: 'vapi_unit_test',
```
s/vapi/volapi/

Call the var `log` instead of `logger` perhaps? Suspect you just copied other test files here tho.

misterdjules replied:
Will fix.

trentm commented:
Done my first review. :)

misterdjules commented:
@trentm Addressed most of your comments except for:

* https://github.com/joyent/node-sdc-clients/pull/8#discussion_r59661691: we still need to have that discussion.
* https://github.com/joyent/node-sdc-clients/pull/8#discussion_r59662138: it's actually not clear why the API would need that now, so I removed it for now, we can also discuss that and re-add it later if that's relevant.

misterdjules commented:
Work on this has been happening in the `tritonnfs` feature branch, and will be reviewed with Gerrit at cr.joyent.us.
